rs/migration_canister/src/validation.rs

```diff
@@ -26,52 +27,72 @@ pub async fn validate_request(
     target: Principal,
     caller: Principal,
 ) -> Result<Request, ValidationError> {
-    // 1. Is any of these canisters already in a migration process?
+    // We first check if the caller is authorized (i.e.,
+    // if the caller is a controller of both the source and target)
+    // before acquiring locks for the source and target
+    // to prevent unauthorized callers from acquiring the lock
+    // and blocking authorized callers from performing canister migration.
+
+    // 1. Is the caller controller of the source? This call also fails if we are not controller.
+    let source_status = canister_status(source)
+        .await
+        .into_result(&format!("Call to management canister (`canister_status`) failed. Ensure that the canister {} is the expected source and try again later.", source))?;
+    if !source_status.settings.controllers.contains(&caller) {
+        return Err(ValidationError::CallerNotController { canister: source });
+    }
+    // 2. Is the caller controller of the target? This call also fails if we are not controller.
+    let target_status = canister_status(target)
+        .await
+        .into_result(&format!("Call to management canister (`canister_status`) failed. Ensure that the canister {} is the expected target and try again later.", target))?;
+    if !target_status.settings.controllers.contains(&caller) {
+        return Err(ValidationError::CallerNotController { canister: target });
+    }
+
+    // 3. The source must not be equal to the target.
+    if source == target {
+        return Err(ValidationError::SameSubnet);
+    }
+
+    // Now we can acquire the locks
+    // to prevent reentrancy bugs across asynchronous calls
+    // while validating the source and target.
+    let Ok(_source_guard) = CanisterGuard::new(source) else {
+        return Err(ValidationError::ValidationInProgress { canister: source });
+    };
+    let Ok(_target_guard) = CanisterGuard::new(target) else {
+        return Err(ValidationError::ValidationInProgress { canister: target });
+    };
+
+    // 4. Is any of these canisters already in a migration process?
     for request in list_by(|_| true) {
         if let Some(id) = request.request().affects_canister(source, target) {
             return Err(ValidationError::MigrationInProgress { canister: id });
         }
     }
-    // 2. Does the source canister exist?
+    // 5. Are the source and target on the same subnet?
     let source_subnet = get_subnet_for_canister(source).await?;
-    // 3. Does the target canister exist?
     let target_subnet = get_subnet_for_canister(target).await?;
```

Comment:
These calls do not really check if the canister exists, they only check if the canister ID belongs to an existing subnet. Now we make `canister_status` calls before, they can no longer fail, but it is still more robust to not unwrap here.